# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/sandbox/chromium/base/location.h
# Commit: 370c40e49230
# Full Hash: 370c40e4923073a87845b57c1f4851bacd9663da
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-12 21:42:35
# Regressor Bug: 1552160
# File Overlap Count: 3
# Description:
#   Bug 1552160 Part 1: Roll-up of chromium sandbox update and mozilla patches to get a running browser. r=jld,aklotz,tjr,bobowen
#   
#   This updates security/sandbox/chromium/ files to chromium commit 84108231f6e6e0772fb9a4643679ce76aa771e67.
#   
#   Existing and new patches applied from security/sandbox/chromium-shim/patches/with_update/ to give a compiling and mostly working browser.
# ==============================================================================

diff -r bd5ecd72f76c -r 370c40e49230 security/sandbox/chromium/base/location.h
--- a/security/sandbox/chromium/base/location.h	Wed Jun 12 21:55:59 2019 +1200
+++ b/security/sandbox/chromium/base/location.h	Wed Jun 12 11:10:48 2019 +0100
@@ -11,14 +11,23 @@
 #include <string>
 
 #include "base/base_export.h"
+#include "base/debug/debugging_buildflags.h"
 #include "base/hash.h"
 
-namespace tracked_objects {
+namespace base {
 
 // Location provides basic info where of an object was constructed, or was
 // significantly brought to life.
 class BASE_EXPORT Location {
  public:
+  Location();
+  Location(const Location& other);
+
+  // Only initializes the file name and program counter, the source information
+  // will be null for the strings, and -1 for the line number.
+  // TODO(http://crbug.com/760702) remove file name from this constructor.
+  Location(const char* file_name, const void* program_counter);
+
   // Constructor should be called with a long-lived char*, such as __FILE__.
   // It assumes the provided value will persist as a global constant, and it
   // will not make a copy of it.
@@ -27,84 +36,82 @@
            int line_number,
            const void* program_counter);
 
-  // Provide a default constructor for easy of debugging.
-  Location();
-
-  // Copy constructor.
-  Location(const Location& other);
-
-  // Comparator for hash map insertion.
-  // No need to use |function_name_| since the other two fields uniquely
-  // identify this location.
+  // Comparator for hash map insertion. The program counter should uniquely
+  // identify a location.
   bool operator==(const Location& other) const {
-    return line_number_ == other.line_number_ &&
-           file_name_ == other.file_name_;
+    return program_counter_ == other.program_counter_;
   }
 
-  const char* function_name()   const { return function_name_; }
-  const char* file_name()       const { return file_name_; }
-  int line_number()             const { return line_number_; }
+  // Returns true if there is source code location info. If this is false,
+  // the Location object only contains a program counter or is
+  // default-initialized (the program counter is also null).
+  bool has_source_info() const { return function_name_ && file_name_; }
+
+  // Will be nullptr for default initialized Location objects and when source
+  // names are disabled.
+  const char* function_name() const { return function_name_; }
+
+  // Will be nullptr for default initialized Location objects and when source
+  // names are disabled.
+  const char* file_name() const { return file_name_; }
+
+  // Will be -1 for default initialized Location objects and when source names
+  // are disabled.
+  int line_number() const { return line_number_; }
+
+  // The address of the code generating this Location object. Should always be
+  // valid except for default initialized Location objects, which will be
+  // nullptr.
   const void* program_counter() const { return program_counter_; }
 
+  // Converts to the most user-readable form possible. If function and filename
+  // are not available, this will return "pc:<hex address>".
   std::string ToString() const;
 
-  // Hash operator for hash maps.
-  struct Hash {
-    size_t operator()(const Location& location) const {
-      // Compute the hash value using file name pointer and line number.
-      // No need to use |function_name_| since the other two fields uniquely
-      // identify this location.
-
-      // The file name will always be uniquely identified by its pointer since
-      // it comes from __FILE__, so no need to check the contents of the string.
-      // See the definition of FROM_HERE in location.h, and how it is used
-      // elsewhere.
-      return base::HashInts(reinterpret_cast<uintptr_t>(location.file_name()),
-                            location.line_number());
-    }
-  };
-
-  // Translate the some of the state in this instance into a human readable
-  // string with HTML characters in the function names escaped, and append that
-  // string to |output|.  Inclusion of the file_name_ and function_name_ are
-  // optional, and controlled by the boolean arguments.
-  void Write(bool display_filename, bool display_function_name,
-             std::string* output) const;
-
-  // Write function_name_ in HTML with '<' and '>' properly encoded.
-  void WriteFunctionName(std::string* output) const;
+  static Location CreateFromHere(const char* file_name);
+  static Location CreateFromHere(const char* function_name,
+                                 const char* file_name,
+                                 int line_number);
 
  private:
-  const char* function_name_;
-  const char* file_name_;
-  int line_number_;
-  const void* program_counter_;
-};
-
-// A "snapshotted" representation of the Location class that can safely be
-// passed across process boundaries.
-struct BASE_EXPORT LocationSnapshot {
-  // The default constructor is exposed to support the IPC serialization macros.
-  LocationSnapshot();
-  explicit LocationSnapshot(const tracked_objects::Location& location);
-  ~LocationSnapshot();
-
-  std::string file_name;
-  std::string function_name;
-  int line_number;
+  const char* function_name_ = nullptr;
+  const char* file_name_ = nullptr;
+  int line_number_ = -1;
+  const void* program_counter_ = nullptr;
 };
 
 BASE_EXPORT const void* GetProgramCounter();
 
-// Define a macro to record the current source location.
+// The macros defined here will expand to the current function.
+#if BUILDFLAG(ENABLE_LOCATION_SOURCE)
+
+// Full source information should be included.
 #define FROM_HERE FROM_HERE_WITH_EXPLICIT_FUNCTION(__func__)
+#define FROM_HERE_WITH_EXPLICIT_FUNCTION(function_name) \
+  ::base::Location::CreateFromHere(function_name, __FILE__, __LINE__)
+
+#else
+
+// TODO(http://crbug.com/760702) remove the __FILE__ argument from these calls.
+#define FROM_HERE ::base::Location::CreateFromHere(__FILE__)
+#define FROM_HERE_WITH_EXPLICIT_FUNCTION(function_name) \
+  ::base::Location::CreateFromHere(function_name, __FILE__, -1)
 
-#define FROM_HERE_WITH_EXPLICIT_FUNCTION(function_name)                        \
-    ::tracked_objects::Location(function_name,                                 \
-                                __FILE__,                                      \
-                                __LINE__,                                      \
-                                ::tracked_objects::GetProgramCounter())
+#endif
+
+}  // namespace base
+
+namespace std {
 
-}  // namespace tracked_objects
+// Specialization for using Location in hash tables.
+template <>
+struct hash<::base::Location> {
+  std::size_t operator()(const ::base::Location& loc) const {
+    const void* program_counter = loc.program_counter();
+    return base::Hash(&program_counter, sizeof(void*));
+  }
+};
+
+}  // namespace std
 
 #endif  // BASE_LOCATION_H_